crates/ruff_python_formatter/README.md

```diff
@@ -0,0 +1,306 @@
+like the following, where moving the `# noqa` to the end of the line causes it to suppress errors
+on both `first()` and `second()`:
+
+```python
+# Input
+[
+    first(),  # noqa
+    second()
+]
+
+# Black
+[first(), second()]  # noqa
+
+# Ruff
+[
+    first(),  # noqa
+    second(),
+]
+```
+
+### Line width vs. line length
+
+Ruff uses the unicode width of a line to determine if a line fits. Black's stable style uses
+character width, while Black's preview style uses unicode width for strings ([#3445](https://github.com/psf/black/pull/3445)),
+and character width for all other tokens.
+
+### Walruses in slice expressions
+
+Black avoids inserting space around `:=` operators within slices. For example, the following adheres
+to Black stable style:
+
+```python
+# Input
+x[y:=1]
+
+# Black
+x[y:=1]
+```
+
+Ruff will instead add space around the `:=` operator:
+
+```python
+# Input
+x[y:=1]
+
+# Ruff
+x[y := 1]
+```
+
+This will likely be incorporated into Black's preview style ([#3823](https://github.com/psf/black/pull/3823)).
+
+### `global` and `nonlocal` names are broken across multiple lines by continuations
+
+If a `global` or `nonlocal` statement includes multiple names, and exceeds the configured line
+width, Ruff will break them across multiple lines using continuations:
+
+```python
+# Input
+global analyze_featuremap_layer, analyze_featuremapcompression_layer, analyze_latencies_post, analyze_motions_layer, analyze_size_model
+
+# Ruff
+global \
+    analyze_featuremap_layer, \
+    analyze_featuremapcompression_layer, \
+    analyze_latencies_post, \
+    analyze_motions_layer, \
+    analyze_size_model
+```
+
+### Newlines are inserted after single-quoted docstrings
+
+Black typically enforces a single newline after a class docstring. However, it does not apply such
+formatting if the docstring is single-quoted rather than triple-quoted, while Ruff enforces a
+single newline in both cases:
+
+```python
+# Input
+class IntFromGeom(GEOSFuncFactory):
+    """Argument is a geometry, return type is an integer."""
+    argtypes = [GEOM_PTR]
+    restype = c_int
+    errcheck = staticmethod(check_minus_one)
+
+# Black
+class IntFromGeom(GEOSFuncFactory):
+    """Argument is a geometry, return type is an integer."""
+    argtypes = [GEOM_PTR]
+    restype = c_int
+    errcheck = staticmethod(check_minus_one)
+
+# Ruff
+class IntFromGeom(GEOSFuncFactory):
+    """Argument is a geometry, return type is an integer."""
+
+    argtypes = [GEOM_PTR]
+    restype = c_int
+    errcheck = staticmethod(check_minus_one)
+```
+
+### Trailing own-line comments on imports are not moved to the next line
+
+Black enforces a single empty line between an import and a trailing own-line comment. Ruff leaves
+such comments in-place:
+
+```python
+# Input
+import os
+# comment
+
+import sys
+
+# Black
+import os
+
+# comment
+
+import sys
+
+# Ruff
+import os
+# comment
+
+import sys
+```
```

Comment:
Is the goal to document all deviations or the most prominent once? There are e.g. some more string formatting deviations https://www.notion.so/astral-sh/Black-Deviations-65977bec23b64aee9d28cc1787badfdb?pvs=4#df068d4a5638465ab95e944f45daac55